parsing/parsetree.mli

```diff
@@ -215,10 +215,12 @@ and pattern_desc =
 
            Invariant: n >= 2
         *)
-  | Ppat_construct of Longident.t loc * pattern option
-        (* C                None
-           C P              Some P
-           C (P1, ..., Pn)  Some (Ppat_tuple [P1; ...; Pn])
+  | Ppat_construct of
+      Longident.t loc * (pattern * (string loc list * core_type) option) option
+        (* C                    None
+           C P                  Some (P, None)
+           C (P1, ..., Pn)      Some (Ppat_tuple [P1; ...; Pn], None)
+           C (type a b) (P : T) Some (P, Some ([a; b], T))
```

Comment:
Sorry, I'm late in the discussion, but shouldn't we encode a regular Ppat_constraint to keep the core_type, i.e. just use:

````ocaml
  Ppat_construct of loc * (pattern * string loc list) option
````

(with an empty list when there is no `(type ...)`)  Requiring to have the inner pattern be a Ppat_constraint when the list is not empty is ok, imo.

I'm concerned that the current representation changes the way an existing constraint would be represented when one adds a type binding (`C (P : T)` --> `C (type a) (P : T')`).

Can the type-checked behave differently on `C (type a) (P : T)` and `C (type a) ((P : T) : T)`?

Btw, is it an error to bind type names which are not used in the pattern?

Reply:
I thought about suggesting such nesting, but:
- Currently the annotation is interpreted in the context of the type variables in a very specific way, so it makes sense to keep the two pieces together at the constructor site.
- Even with the representation you propose, my understanding is that the type-checker would have to lookup the `Tpat_constraint` node and use its payload specifically (not go through the usual `Tpat_constraint` path in this case), so you would not get the same-behavior guarantees that you are asking about.

The handling of existential variables is very specific because, morally, their actual binding is computed by unifying the annotation with the constructor return type, and then making them rigid for the rest of the type-checking. This is not the behavior you would get if you checked the whole subpattern before rigidifying (I think that behavior would make sense and we discussed it earlier, but @garrigue says it is harder to implement), nor of course the behavior you would get by rigidifying before checking the annotation or using flexible variables.

Reply:
Note: the previous discussion I was referring to is starting (approximatively) at https://github.com/ocaml/ocaml/issues/7074#issuecomment-630847726 .

Reply:
Thanks @gasche for the pointer.

 - If the type-checking is not an extension of what happens for `C (P : T)`, reusing the same concrete syntax can really be misleading, no?

 - Independently of what the type checker does, having different representation for the T in `C (type a) (P : T)` and in `C (P : T)` is confusing for people writing syntactic tools.  I'm not shocked that, short of having a cleaner solution, the type-checker does something special with Ppat_constraint on top of the pattern when existential type names are bound.  We might be able to change the behavior of the type-checker later, while keeping the same syntax and Parsetree representation.

Reply:
> If the type-checking is not an extension of what happens for C (P : T), reusing the same concrete syntax can really be misleading, no?

I think that the two readings of `C (P : T)` (as a constraint, and as a return-type-annotated-constructor pattern with no existential variables bound) do have the same meaning. What I meant is that if we think of the type-checking of a pattern `P` as generating an inference constraint (or some other "result", for example in an non-inferred world taking an scrutinee type and returning types for the bound variables), then the result of type-checking of `C (type a...) (P : T)` cannot be computed in a modular lay, as an action on the result of the type-checking of `(P : T)`.

> Independently of what the type checker does, having different representation for the T in C (type a) (P : T) and in C (P : T) is confusing for people writing syntactic tools. 

This is a good point.

Reply:
Answering @alainfrisch 's concerns:
* Except for the special handling of variables introduced by `(type a ...)`, the code is supposed to do exactly the same thing as `Tpat_constraint`.
* In particular, `C (type a) (P : T)` and `C (type a) ((P : T) : T)` should be strictly equivalent. I.e., `a` is bound by the external type annotation, and reusing in the internal one should give the same type.
* It is not an error to introduce a type name without binding it. It just has the same meaning as a usual `(type a)`.
* However, one must bind all existentials arising from the constructor, with distinct names, and only once.

Of course, the first part is only true if there is no bug in the code, which I hope is the case, but the complexity of `Tpat_construct` for GADT constructors makes the comparison non-trivial.
For the last point, here are some examples:
```ocaml
# type dyn2 = Dyn : 'a * 'b -> dyn2;;
# function Dyn (type a b) (_ : a * b) -> ();;
- : dyn2 -> unit = <fun>
# function Dyn (type a) (_ : a * _) -> ();;
Error: This type does not bind all existentials in the constructor:
         type a. a * 'a
# function Dyn (type a) (_ : a * a) -> ();;
Error: This type does not bind all existentials in the constructor:
         type a. a * a
# type dynp = Dynp : 'a * 'a -> dynp;;
type dynp = Dynp : 'a * 'a -> dynp
# function Dynp (type a b) (_ : a * b) -> ();;
Error: This pattern matches values of type a * b
       but a pattern was expected which matches values of type a * a
       Type b is not compatible with type a 
```

Reply:
>  then the result of type-checking of C (type a...) (P : T) cannot be computed in a modular lay, as an action on the result of the type-checking of (P : T).

But that's not something unexpected for a binding construction, right?

If I extrapolate @garrigue's answer, the code currently used for `C (type ...) (P : T)` should do the same thing when applied to an empty list of binders as the existing `C (P : T)` (modulo bugs).  It's even more tempting to use the representation I proposed.

Note that expecting a specific root structure in the Parsetree is not something unheard of : n-ary constructors are parsed and represented in the Parsetree as tuple, and the type-checker reacts in a special way to those Ppat_tuple/Pexp_tuple under n-ary constructor (this is even less local than what I propose, which depends on the presence of a non-empty of binders, i.e. a local syntactic property).

> This type does not bind all existentials in the constructor:

Why is this rejected?  Could those remaining existentials be treated as they are today, when there is no binding?

Reply:
Personally, I prefer the current representation. The proposed change adds yet another hidden invariant to the parsetree, making it harder to reliably generate valid parsetrees in ppxes. It's similar to the current encoding of `let x : 'a. 'a -> 'a = ...` which has been a persistent cause of pain to ppx authors.

Reply:
I think that Alain's proposal would make a lot of sense if we had a modular desugaring of `C (type a ..) (P : T)` into `(type a ...) P` (which introduces open existential types with can only be unified with distinct pure existentials within P and are opaque to the outside) and `C P` and `(P : T)`. But we are not currently able to implement this richer (and I think nicer) semantics today.

If we had good reasons to believe that we would be able to have such open-existential-types in the medium term, then it would make sense to argue for this representation now (even if it is somewhat of a lie today). But so far I was the only one to advocate for this extended feature, and I'm not expert enough to be confident that it even makes sense.

Reply:
>  The proposed change adds yet another hidden invariant to the parsetree, making it harder to reliably generate valid parsetrees in ppxes.

It's a matter of perspective.  I tend to think that `C (type a ...) P` is a type error when P is not a constraint, not a syntax error, so it's really not a matter of generating valid or invalid parstrees in ppx, or having hidden invariants (unless you consider all type-checking errors as hidden invariants!).  Moreover, it would be possible to later support putting constraints deeper in the pattern, and this would just be an improvement of the type-checker, with no need to change the syntax (e.g. it should be straightforward to accept `C (type a) ((P1 : t1), (P2 : t2))`, understood as `C (type a) ((P1, P2) : t1 * t2)`).

Reply:
If you going to take that perspective then I really think the parser needs to accept things like `C (type a) 4` and we should emit the error from the type-checker instead.

Reply:
Is there a possibility that we could postpone this discussion for a later PR, and merge this one with this question delayed? If I was in @garrigue's stead I would start getting slightly irritated by the back-and-forth.

Reply:
> If you're going to take that perspective then I really think the parser needs to accept things like C (type a) 4 and we should emit the error from the type-checker instead.

Yes, that's what I'm proposing.  I think it might give better error messages, btw.

> Is there a possibility that we could postpone this discussion for a later PR, and merge this one with this question delayed? If I was in @garrigue's stead I would start getting slightly irritated by the back-and-forth.

I did not follow the latest developments in the ppx ecosystem, but last time I checked, there was relatively high cost associated to changing the Parsetree definition.  If we can avoid having to adapt tooling for an intermediate version, that might be better.

But I take your point.  I came late to the discussion, so I'm not in a good position to block it if everyone else is ok with the current state.

Reply:
I think that your argument does have merit (@lpw25 seems against and @garrigue has not spoken yet, but this is something I have considered as well and would find somewhat natural). We should take the time to discuss it, but maybe in a non-blocking way indeed.

Reply:
(When I say "merge this one", I'm not actually sure that all pending questions have been resolved, this is for @lpw25 to confirm. So it may not be the unique blocker *yet*.)

Reply:
I'm fine with either the current version or the version where the check is completely moved to the type-checker and the parser is happy to parse patterns with existential binders and no type constraints. I dislike any version where what the type can represent and what the parser can produce are not the same.

Reply:
I just realize that it is wrong to say that
```
C (type a) (pats : ty)
```
is equivalent to
```
C (pats : ty)
```
when `a` occurs nowhere: if `C` has arity 2 or more, `pats` must syntactically be a list of patterns, and the latter would not be typable. So, sorry, this is really a new syntax.
@gasche indeed suggested that it would be nicer to allow type constraints on each argument, but I answered that it would be difficult to implement it cleanly.
I will still give it a try, just to see how it would work.

Reply:
So, IIUC, for a pattern with 2 arguments:
  - `C _` is allowed but `C (type a) _` is rejected(?).
  - `C (type a) ((x, y) : t)` can be allowed (for a correct type) (?) but `C ((x, y) : t)` is always rejected.

Is that right? Are there other differences?

The situation with n-ary arguments was already quite confusing, I fear this is getting even worse...

Reply:
Not quite correct.
* `C _` is allowed, but `C (type a) _` is rejected by the type checker (not the parser) independently of the number of arguments. The `(type a)` notation inside patterns is restricted to existential variables, and requires to bind all of them, which you cannot do without a type annotation, hence the rejection.
* Annotations are always allowed for multiary types (the internal parenthesis are optional).
```ocaml
type ('a, 'b) pair = Pair of 'a * 'b;;
type t = int * bool;;
let f = function Pair ((x,y) : t) -> x;;
val f : (int, bool) pair -> int = <fun>
type closure = Clos : ('a -> int) * 'a -> closure;;
type 'a app = ('a -> int) * 'a;;
let f = function Clos (type a) ((x,y) : a app) -> x y;;
val f : closure -> int = <fun>
```
The result looks more regular to me.